### ISSUE TITLE
perf(morphology): the default `engine="unfold"` is 8x slower than `engine="convolution"` on MPS (identical output), and the docstring's speed claim is inverted on CPU

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4525]
## Problem

`dilation`, `erosion`, `opening`, `closing`, `gradient`, `top_hat`, `bottom_hat` default to `engine="unfold"`: the image is unfolded into a 6-D `(B, C, H, W, kh, kw)` window tensor and reduced with `max`/`min` over the window axes. On MPS that reduction is a weak kernel (a reduction whose innermost dim is 3–7 wide runs 2–5x slower than the CPU; details and standalone repros in #4522), so the whole morphology family is 1.5–2.1x **slower** than the M1 CPU at every shape up to 1024², while `engine="convolution"` is 8x faster on MPS and gives identical output for flat structuring elements.

| op, 256², k7 | CPU unfold | CPU convolution | MPS unfold | MPS convolution | max abs diff between engines |
|---|---:|---:|---:|---:|---:|
| `dilation` | 37.7 ms | 52.0 ms | 58.7 ms | **7.2 ms** | 0.0 |
| `erosion` | 34.5 ms | 38.1 ms | 56.2 ms | **7.0 ms** | 0.0 |
| `dilation`, 512², k7 | 141.5 ms | 214.0 ms | 221.6 ms | **23.1 ms** | 0.0 |

So the CPU prefers `unfold` and accelerators prefer `convolution`, and the docstring says the opposite of the CPU result: "convolution is faster and less memory hungry, and unfold is more stable numerically".

## Proposal

- Pick the engine by device when the caller does not specify one (`unfold` on CPU, `convolution` on CUDA/MPS), or
- keep the default but document which engine wins where, with numbers.

Related: #2948 / #3001 (memory of the unfold engine at large kernels), #4521, #4522.

Repro:

```python
import torch, torch.utils.benchmark as bench, kornia.morphology as KM
def us(fn, dev):
    s = torch.mps.synchronize if dev == "mps" else (lambda: None); fn(); s()
    return bench.Timer(stmt="fn(); s()", globals={"fn": fn, "s": s}).blocked_autorange(min_run_time=0.4).median * 1e6
for dev in ("cpu", "mps"):
    im = torch.rand(1, 3, 256, 256, device=dev); ker = torch.ones(7, 7, device=dev)
    a, b = KM.dilation(im, ker, engine="unfold"), KM.dilation(im, ker, engine="convolution")
    print(dev, us(lambda: KM.dilation(im, ker, engine="unfold"), dev), us(lambda: KM.dilation(im, ker, engine="convolution"), dev), (a - b).abs().max().item())
```

Measured on torch nightly 2.15.0.dev20260913, Apple M1 8 GB, macOS 26.5.1, kornia `633951bd7`, `PYTORCH_ENABLE_MPS_FALLBACK` unset, `torch.utils.benchmark` medians with `mps.synchronize` inside the timed region, fresh process per op. Split out of the MPS-audit tracking issue #4521; the PyTorch-side kernel findings are in #4522.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
